hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java

```diff
@@ -1071,4 +1071,33 @@ static boolean isCompositeId(XClass entityClass, XProperty idProperty) {
 	public static boolean isDefault(XClass clazz, MetadataBuildingContext context) {
 		return context.getBootstrapContext().getReflectionManager().equals( clazz, void.class );
 	}
+
+	public static void checkMappedByType(
+			String mappedBy,
+			Value targetValue,
+			String propertyName,
+			PropertyHolder propertyHolder) {
+		final ToOne toOne;
+		if ( targetValue instanceof Collection ) {
+			toOne = (ToOne) ( (Collection) targetValue ).getElement();
+		}
+		else {
+			toOne = (ToOne) targetValue;
+		}
+		final String referencedEntityName = toOne.getReferencedEntityName();
+		PersistentClass referencedClass = propertyHolder.getPersistentClass();
+		while ( referencedClass != null ) {
+			if ( referencedClass.getEntityName().equals( referencedEntityName ) ) {
+				return;
+			}
+			else {
+				referencedClass = referencedClass.getSuperclass();
+			}
+		}
+		throw new AnnotationException(
+				"Collection '" + qualify( propertyHolder.getPath(), propertyName )
+						+ "' is 'mappedBy' a property named '" + mappedBy
+						+ "' which references the wrong entity type '" + referencedEntityName + "'"
```

Comment:
The message doesn't fit for ToOne attributes. Would be cool if you could adapt that a bit. Also, would be nice if you could add the expected entity name to the error message.